### PR TITLE
Adding deep learning to nussl

### LINKED ISSUE #123: Onboarding users new to source separation
Source separation is still a relatively esoteric research field, so it would be nice to have materials for users who have no idea how to use source separation. We should implement this in two ways:

- [ ]  A how-to guide in the documentation that gives a broad overview of algorithms. E.g., "If your signal has X and is Y, try algorithm A, which has such and such assumptions, or algorithm B."  A flow chart would also be nice.

- [ ]  Additionally, docs for individual algorithms should clearly outline the assumptions the algorithms make, and what the strengths and weaknesses are for each algorithm.

### LINKED ISSUE #133: module 'nussl' has no attribute 'AudioSignal'
#### Description

When I import nussl and run:
```python
history = nussl.AudioSignal('HistoryRepeatingPropellerHeads.wav')
```

It gives me the error in the title.

#### Steps/Code to Reproduce

see above

#### Expected Results

No error occurs

#### Actual Results

This output:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'nussl' has no attribute 'AudioSignal'
```

#### Versions

```
import platform; print(platform.platform())
Darwin-18.2.0-x86_64-i386-64bit
>>> import sys; print("Python", sys.version)
Python 3.7.1 (default, Nov  6 2018, 18:46:03) 
[Clang 10.0.0 (clang-1000.11.45.5)]
>>> import numpy; print("NumPy", numpy.__version__)
NumPy 1.15.4
>>> import scipy; print("SciPy", scipy.__version__)
SciPy 1.2.0
>>> import sklearn; print("Scikit-Learn", sklearn.__version__)
Scikit-Learn 0.20.2
>>> import librosa; print("librosa", librosa.__version__)
librosa 0.6.2
>>> import nussl; print("nussl", nussl.__version__)
```

### PR DESCRIPTION
**Pull request by pseeth**
#### Reference Issue

Fixes #133, #143, #142, #141 #140, #139, #132, #130, #135, #127, #115, #85, #145 

#### What does this implement/fix? Explain your changes.
The most significant contribution in this pull request is the integration of deep learning into nussl. Deep learning is the current state-of-the-art for source separation and it is currently not in nussl. This pull request integrates both training and inference code for three deep learning approaches and is extensibly for many more. It includes Deep Clustering, Mask Inference, and Chimera networks. Along with this, it also fixes a number of issues listed above. Finally, it updates nussl to work with (and only with) Python 3.6 and above. Python 2 is deprecated by the end of the year and Python 3.5 and below are officially in maintenance mode. nussl should work with the latest Python.

#### Any other comments?
This is a new pull request to merge things into development instead of master. Changes get staged in nussl's dev branch and then merged into master along with a pip install.

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1,5 @@
+include nussl/deep/config/*.json
+include nussl/deep/config/defaults/*.json
+include nussl/deep/config/defaults/metadata/*.json
+include nussl/deep/config/defaults/models/*.json
+include nussl/deep/config/defaults//models/metadata/*.json
```
Comment (nathanshelly) [inline MANIFEST.in]: any reason for the double slash? - `defaults//models`

File: README.md [hunk 2]
```diff
@@ -45,26 +48,37 @@ Spatialization algorithms:
 * Degenerate Unmixing Estimation Technique (DUET)
 * PROJET 
 
-Median filtering algorithms:
+Repetition algorithms:
```
Comment (nathanshelly) [inline README.md]: could be worth making these things headers

File: README.md [hunk 3]
```diff
@@ -17,12 +19,13 @@ Please see the issues page before contacting the authors.
 Branch Layout
```
Comment (nathanshelly) [inline README.md]: could make header instead of manual -------

File: README.md [hunk 4]
```diff
@@ -2,12 +2,14 @@
 nussl
```
Comment (nathanshelly) [inline README.md]: would use headers instead of manual separation w/ `=====`

File: README.md [hunk 5]
```diff
@@ -77,62 +91,72 @@ Installation is easy if you have pip (recommended):
...
+Corey Grief ([website](http://music.cs.northwestern.edu/emeritus.php)), 
+Daniel Felix Kim, Ben Kalish
 
 
 Contact
 -------
-Contact Ethan Manilow (ethanmanilow [at] u.northwestern.edu) with any questions or issues. Please look at the
-"issues" page before reporting problems.
+Contact Ethan Manilow (ethanmanilow [at] u [dot] northwestern [dot] edu) with any questions or 
+issues. Please look at the "issues" page before reporting problems.
```
Comment (nathanshelly) [inline README.md]: could link to issues page here for the lazy :)

**Comment by zaptrem** [comment]
Is this still under development?

**Comment by ethman** [comment]
> Is this still under development?

Yup, we're still working on it. We've been working behind the scenes, but we should be rolling something out soon.

**Comment by zaptrem** [comment]
> > Is this still under development?
> 
> Yup, we're still working on it. We've been working behind the scenes, but we should be rolling something out soon.

Cool. Any way to use the changes you've made so far for basic experiments? It seems clustering is the most up-to-date repo but it was completely broken on my system (might have been a setup mistake on my part).

**Comment by joaquim679** [comment]
I am stuck with " _ModuleNotFoundError: No module named 'audio_signal'_ "  any pointers.
I am on the nussl train branch with python 3

===========================================
Traceback (most recent call last):
  File "test_nussl.py", line 1, in <module>
    import nussl
  File "<frozen zipimport>", line 259, in load_module
  File "/media/joaquim/Data/anaconda2/envs/envn3/lib/python3.8/site-packages/nussl-0.1.6-py3.8.egg/nussl/__init__.p
y", line 26, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/media/joaquim/Data/anaconda2/envs/envn3/lib/python3.8/site-packages/nussl-0.1.6-py3.8.egg/nussl/core/__ini
t__.py", line 7, in <module>
ModuleNotFoundError: No module named 'audio_signal'

**Comment by pseeth** [comment]
Hi @zaptrem, we are working on getting everything up and running properly. Should be done soon! I can ping you as soon as it is done.

**Comment by zaptrem** [comment]
Sounds great, thank you!

**Comment by joaquim679** [comment]
All the best to the team !!

**Comment by pseeth** [comment]
Here's your ping, as promised, albeit a little late!

**Comment by joaquim7** [comment]
Thank you

**Pull request closed without merging**